runtime/src/bank.rs

```diff
@@ -395,42 +445,97 @@ impl CachedExecutors {
             max,
             current_epoch,
             executors: HashMap::new(),
+            stats: executor_cache::Stats::default(),
         }
     }
+
     fn get(&self, pubkey: &Pubkey) -> Option<Arc<dyn Executor>> {
         self.executors.get(pubkey).map(|entry| {
             entry.epoch_count.fetch_add(1, Relaxed);
             entry.executor.clone()
         })
     }
+
     fn put(&mut self, pubkey: &Pubkey, executor: Arc<dyn Executor>) {
-        if !self.executors.contains_key(pubkey) && self.executors.len() >= self.max {
+        let entry = if let Some(mut entry) = self.executors.remove(pubkey) {
+            saturating_add_assign!(self.stats.hits, 1);
+            entry.executor = executor;
+            entry
+        } else {
+            saturating_add_assign!(self.stats.misses, 1);
             let mut least = u64::MAX;
             let default_key = Pubkey::default();
             let mut least_key = &default_key;
 
-            for (key, entry) in self.executors.iter() {
-                let count = entry.prev_epoch_count + entry.epoch_count.load(Relaxed);
-                if count < least {
-                    least = count;
-                    least_key = key;
-                }
-            }
+            let mut count_sum = 0u64;
+            let mut counts = self
+                .executors
+                .iter()
+                .map(|(key, entry)| {
+                    let count = entry.prev_epoch_count + entry.epoch_count.load(Relaxed);
+                    if count < least {
+                        least = count;
+                        least_key = key;
+                    }
+                    saturating_add_assign!(count_sum, count);
+                    count
+                })
+                .collect::<Vec<_>>();
+
             let least_key = *least_key;
-            let _ = self.executors.remove(&least_key);
-        }
-        let _ = self.executors.insert(
-            *pubkey,
+
+            if self.executors.len() >= self.max {
+                let _ = self.executors.remove(&least_key);
+                self.stats
+                    .evictions
+                    .entry(least_key)
+                    .and_modify(|c| saturating_add_assign!(*c, 1))
+                    .or_insert(1);
+            }
+
+            counts.sort_unstable();
+            let primer_count = Self::get_primer_count(&counts, count_sum);
+
             CachedExecutorsEntry {
                 prev_epoch_count: 0,
-                epoch_count: AtomicU64::new(0),
+                epoch_count: AtomicU64::new(primer_count),
                 executor,
-            },
-        );
+            }
+        };
+        let _ = self.executors.insert(*pubkey, entry);
     }
+
     fn remove(&mut self, pubkey: &Pubkey) {
         let _ = self.executors.remove(pubkey);
     }
+
+    fn get_primer_count_upper_bound(counts: &[u64], count_sum: u64) -> u64 {
```

Comment:
This is precisely what I wanted to avoid.
Both, measuring relative to the max entry or the sum of all entries is really skewed in pareto distributions. Instead one has to select the entry at a relative rank.

Reply:
Hrm, right.  Got it.  Thanks for the simulation, that helps a lot!  I guess 2am Trent wasn't visualizing the distribution well.

Any ideas how we can reconcile this with over-priming early additions to the cache?  Maybe a `n / max` multiplier?

Reply:
It is like the difference between the average and the median.
I would say we redo this entirely, instead of making it more messy, because:
- We can't guess a priming value without proper ranking / sorting, and that would require an acceleration structure
- The counts should be cached separately of the executors, so that way more counters can be retained (maybe just store one persistent usage counter per program)
- The proposal by @behzadnouri makes more sense in my mind as that takes the median, the only question is how to randomize it

I know this might look like over-engineering, but adjusting this PR is a similar amount of work I guess.

Reply:
> The counts should be cached separately of the executors, so that way more counters can be retained

https://github.com/solana-labs/solana/pull/22364 implements above on top of https://github.com/solana-labs/solana/pull/22344